File: src/openfl/filesystem/File.nim
import std/os

type
  File* = ref object
    nativePath*: string

proc newFile*(path: string = ""): File =
  return File(nativePath: path)

proc exists*(self: File): bool =
  return fileExists(self.nativePath) or dirExists(self.nativePath)

proc isDirectory*(self: File): bool =
  return dirExists(self.nativePath)

proc resolvePath*(self: File, path: string): File =
  return newFile(joinPath(self.nativePath, path))

var applicationDirectory*: File = newFile(getAppDir())
var applicationStorageDirectory*: File = newFile(getAppDir())
var desktopDirectory*: File = newFile(getHomeDir() / "Desktop")
var documentsDirectory*: File = newFile(getHomeDir() / "Documents")
var userDirectory*: File = newFile(getHomeDir())

File: kfl/src/main/kotlin/openfl/filesystem/File.kt
package openfl.filesystem

import openfl.events.EventDispatcher
import java.io.File as JFile

class File(path: String? = null) : EventDispatcher() {
    var nativePath: String = path ?: ""
        private set
    var exists: Boolean = false
        get() = JFile(nativePath).exists()
        private set
    var isDirectory: Boolean = false
        get() = JFile(nativePath).isDirectory
        private set

    fun resolvePath(path: String): File {
        val f = JFile(nativePath, path)
        return File(f.absolutePath)
    }

    companion object {
        val applicationDirectory: File get() = File(System.getProperty("user.dir"))
        val applicationStorageDirectory: File get() = File(System.getProperty("user.home") + "/.kfl/storage")
        val desktopDirectory: File get() = File(System.getProperty("user.home") + "/Desktop")
        val documentsDirectory: File get() = File(System.getProperty("user.home") + "/Documents")
        val userDirectory: File get() = File(System.getProperty("user.home"))
    }
}
